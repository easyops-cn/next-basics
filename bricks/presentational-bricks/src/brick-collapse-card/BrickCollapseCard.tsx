import React from "react";
import { Icon as LegacyIcon } from "@ant-design/compatible";
import { Collapse } from "antd";
import { GeneralIcon } from "@next-libs/basic-components";
import { MenuIcon } from "@next-core/brick-types";
import classNames from "classnames";

interface BrickCollapseCardProps {
  title: string;
  expandActiveText?: string;
  expandInactiveText?: string;
  expandActiveIcon?: string;
  expandInactiveIcon?: string;
  isActive?: boolean;
  containerStyle?: any;
  headerStyle?: any;
  contentStyle?: any;
  dataSource?: Record<string, any>;
  hasHeaderSlot?: boolean;
  titleIcon?: MenuIcon | string;
  descriptionList?: descriptionsItemProps[];
  titleWithIconAndDesc?: boolean;
  verticalCenter?: boolean;
}

export interface descriptionsItemProps {
  label: string;
  text: string;
}

const getIconNode = (
  icon: MenuIcon | string,
  className: string
): React.ReactElement => {
  if (typeof icon === "string")
    return <LegacyIcon type={icon} className={className} />;
  if (typeof icon === "object")
    return (
      <span className={className}>
        <GeneralIcon icon={icon} />
      </span>
    );
};

const getExtraNode = (props: BrickCollapseCardProps): React.ReactElement => {
  return (
    <span className="toggleCollapse">
      <span className="expandText">
        {props.isActive ? props.expandActiveText : props.expandInactiveText}
      </span>
      {getIconNode(
        props.isActive ? props.expandActiveIcon : props.expandInactiveIcon,
        "expandIcon"
      )}
    </span>
  );
};

const getDescriptionNode = (
  descriptionList: descriptionsItemProps[]
): React.ReactElement[] => {
  return descriptionList.map((item, i) => (
    <span className="descriptionItem" key={i}>
      <span className="descriptionItemLabel">
        {item.label && item.label + ":"}
      </span>
      <span className="descriptionItemText">{item.text}</span>
    </span>
  ));
};

const getHeaderWithIconAndDescNode = (
  props: BrickCollapseCardProps
): React.ReactElement => {
  return (
    <>
      <div className="headerIcon">{getIconNode(props.titleIcon, "")}</div>
      <div className="headerContent">
        <div className="cardTitle">{props.title}</div>
        <div className="descriptionList">
          {props.descriptionList && getDescriptionNode(props.descriptionList)}
        </div>
      </div>
    </>
  );
};

const getHeaderNode = (props: BrickCollapseCardProps): React.ReactElement => {
  if (props.hasHeaderSlot) return <slot id="headerSlot" name="header" />;
  return (
    <div className="collapseTitle">
      {props.titleWithIconAndDesc
        ? getHeaderWithIconAndDescNode(props)
        : props.title}
    </div>
  );
};

export function BrickCollapseCard(
  props: BrickCollapseCardProps
): React.ReactElement {
  return (
    <Collapse activeKey={props.isActive ? "content" : ""} ghost={true}>
      <Collapse.Panel
        style={props.containerStyle}
        forceRender={true}
        key="content"
        showArrow={false}
        header={getHeaderNode(props)}
        extra={getExtraNode(props)}
      >
        <div
          className={classNames("contentContainer", {
            verticalCenter: props.verticalCenter,
          })}
          style={props.contentStyle}
        >
          <slot id="contentSlot" name="content" />
        </div>
      </Collapse.Panel>
    </Collapse>
  );
}
