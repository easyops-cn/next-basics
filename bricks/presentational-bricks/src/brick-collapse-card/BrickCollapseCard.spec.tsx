import React from "react";
import { mount, shallow } from "enzyme";
import { BrickCollapseCard } from "./BrickCollapseCard";

describe("BrickCollapseCard", () => {
  it("should work when isActive is false", () => {
    const wrapper = shallow(
      <BrickCollapseCard
        title="工具"
        expandActiveText="收起"
        expandInactiveText="展开"
        expandActiveIcon="up"
        expandInactiveIcon="down"
        isActive={false}
        hasHeaderSlot={false}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should work when isActive is true", () => {
    const wrapper = shallow(
      <BrickCollapseCard
        title="Sleep"
        expandActiveText="收起"
        expandInactiveText="工具详情"
        expandActiveIcon="up"
        expandInactiveIcon="down"
        isActive={true}
        hasHeaderSlot={true}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should work", () => {
    const wrapper = mount(
      <BrickCollapseCard
        title="标题"
        expandActiveText="收起"
        expandInactiveText="展开"
        expandActiveIcon="up"
        expandInactiveIcon="down"
        isActive={true}
      />
    );
    expect(wrapper.find(".collapseTitle").text()).toEqual("标题");
    expect(wrapper.find(".expandText").text()).toEqual("收起");
    expect(wrapper.find(".ant-collapse-content-active").length).toEqual(1);
    wrapper.setProps({
      isActive: false,
    });
    wrapper.update();
    expect(wrapper.find(".expandText").text()).toEqual("展开");
    expect(wrapper.find(".ant-collapse-content-inactive").length).toEqual(1);
  });
});
