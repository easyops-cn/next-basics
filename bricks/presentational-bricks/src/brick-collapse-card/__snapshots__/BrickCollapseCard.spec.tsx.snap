// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrickCollapseCard should work when isActive is false 1`] = `
<Collapse
  activeKey=""
  ghost={true}
>
  <CollapsePanel
    extra={
      <span
        className="toggleCollapse"
      >
        <span
          className="expandText"
        >
          展开
        </span>
        <Icon
          className="expandIcon"
          type="down"
        />
      </span>
    }
    forceRender={true}
    header={
      <div
        className="collapseTitle"
      >
        工具
      </div>
    }
    key="content"
    showArrow={false}
  >
    <div
      className="contentContainer"
    >
      <slot
        id="contentSlot"
        name="content"
      />
    </div>
  </CollapsePanel>
</Collapse>
`;

exports[`BrickCollapseCard should work when isActive is true 1`] = `
<Collapse
  activeKey="content"
  ghost={true}
>
  <CollapsePanel
    extra={
      <span
        className="toggleCollapse"
      >
        <span
          className="expandText"
        >
          收起
        </span>
        <Icon
          className="expandIcon"
          type="up"
        />
      </span>
    }
    forceRender={true}
    header={
      <slot
        id="headerSlot"
        name="header"
      />
    }
    key="content"
    showArrow={false}
  >
    <div
      className="contentContainer"
    >
      <slot
        id="contentSlot"
        name="content"
      />
    </div>
  </CollapsePanel>
</Collapse>
`;
